Save style sheet resources with same URL in different files

szewai opened the pull request:
#### 04a136a53bf6b2e2dc118536618fd00b41fc65f3
<pre>
Save style sheet resources with same URL in different files
<a href="https://bugs.webkit.org/show_bug.cgi?id=264520">https://bugs.webkit.org/show_bug.cgi?id=264520</a>
<a href="https://rdar.apple.com/118199671">rdar://118199671</a>

Reviewed by Ryosuke Niwa.

In current implementation, if two link elements have the same url, we only create one file for storing the resource.
However, two link elements corresponds to two CSSStyleSheet objects, which can be modified independently by script. To
ensure the changes from script are captured, now we store content of CSSStyleSheet object in its own file.

API tests: WebArchive.SaveResourcesLink
           WebArchive.SaveResourcesLinksWithSameURL
           WebArchive.SaveResourcesCSSImportRule

* Source/WebCore/css/CSSFontFaceRule.cpp:
(WebCore::CSSFontFaceRule::cssTextWithReplacementURLs const):
* Source/WebCore/css/CSSFontFaceRule.h:
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::cssTextForDeclsAndRules const):
* Source/WebCore/css/CSSImportRule.cpp:
(WebCore::CSSImportRule::cssTextInternal const):
(WebCore::CSSImportRule::cssText const):
(WebCore::CSSImportRule::cssTextWithReplacementURLs const):
(WebCore::CSSImportRule::getChildStyleSheets):
* Source/WebCore/css/CSSImportRule.h:
* Source/WebCore/css/CSSRule.h:
(WebCore::CSSRule::cssTextWithReplacementURLs const):
(WebCore::CSSRule::getChildStyleSheets):
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::cssTextForRules const):
(WebCore::CSSStyleRule::cssTextWithReplacementURLs const):
(WebCore::CSSStyleRule::cssTextForRulesWithReplacementURLs const):
(WebCore::CSSStyleRule::getChildStyleSheets):
* Source/WebCore/css/CSSStyleRule.h:
* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::cssTextWithReplacementURLs):
(WebCore::CSSStyleSheet::getChildStyleSheets):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::MarkupAccumulator):
(WebCore::MarkupAccumulator::appendContentsForNode):
(WebCore::MarkupAccumulator::resolveURLIfNeeded const):
* Source/WebCore/editing/MarkupAccumulator.h:
(WebCore::MarkupAccumulator::MarkupAccumulator):
* Source/WebCore/editing/markup.cpp:
(WebCore::serializeFragment):
* Source/WebCore/editing/markup.h:
(WebCore::serializeFragment):
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::styleSheetContentWithReplacementURLs const): Deleted.
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLStyleElement.cpp:
(WebCore::HTMLStyleElement::textContentWithReplacementURLs const): Deleted.
* Source/WebCore/html/HTMLStyleElement.h:
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
(WebCore::addSubresourcesForCSSStyleSheetsIfNecessary):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: <a href="https://commits.webkit.org/270562@main">https://commits.webkit.org/270562@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ee91beb6f9a04bc9f0a1c6bc4e9faf00432d590

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25757 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4364 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27041 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27857 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23592 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26042 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6123 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1800 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26007 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3276 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22199 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28437 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2901 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23154 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29225 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23511 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23524 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27089 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2916 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1149 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4301 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6199 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3368 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3232 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/css/CSSImportRule.cpp:
```diff
@@ -68,11 +68,10 @@ String CSSImportRule::layerName() const
 }
 
-String CSSImportRule::cssText() const
+String CSSImportRule::cssTextInternal(String urlString) const
```
`const String&`

cdumez commented on Source/WebCore/css/CSSImportRule.cpp:
```diff
@@ -87,10 +86,27 @@ String CSSImportRule::cssText() const
+
+String CSSImportRule::cssTextWithReplacementURLs(const HashMap<String, String>& replacementURLStrings, const HashMap<RefPtr<CSSStyleSheet>, String>& replacementURLStringsForCSSStyleSheet) const
+{
+    if (auto sheet = styleSheet()) {
```
`auto` -> `RefPtr`

cdumez commented on Source/WebCore/css/CSSStyleRule.cpp:
```diff
@@ -313,4 +313,10 @@ CSSRuleList& CSSStyleRule::cssRules() const
 
+void CSSStyleRule::getChildStyleSheets(HashSet<RefPtr<CSSStyleSheet>>& childStyleSheets)
+{
+    for (unsigned index = 0 ; index < nestedRules().size() ; index++)
```
`nestedRules().size() -> length()`
`++index`

szewai replied:
I fixed the `++index` part, but not sure what `nestedRules().size() -> length()` means. `size()` returns size_t so it does not have `length()`?

cdumez replied:
Literally replace `nestedRules().size()` with `length()`. Not sure how I can be clearer.

szewai replied:
oh ok, updated both places

cdumez commented on Source/WebCore/css/CSSStyleSheet.cpp:
```diff
@@ -550,6 +550,18 @@ Ref<StyleSheetContents> CSSStyleSheet::protectedContents()
 
+void CSSStyleSheet::getChildStyleSheets(HashSet<RefPtr<CSSStyleSheet>>& childStyleSheets)
+{
+    auto ruleList = cssRules();
```
`auto` -> `RefPtr` for clarity

cdumez commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -211,11 +213,18 @@ String MarkupAccumulator::serializeNodes(Node& targetNode, SerializedNodes root,
 
     auto& styleElement = downcast<HTMLStyleElement>(targetNode);
-    result.append(styleElement.textContentWithReplacementURLs(m_replacementURLStrings));
+    auto* cssStyleSheet = styleElement.sheet();
```
`auto*` -> `RefPtr`

cdumez commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -304,11 +313,19 @@ void MarkupAccumulator::serializeNodesWithNamespaces(Node& targetNode, Serialize
 String MarkupAccumulator::resolveURLIfNeeded(const Element& element, const String& urlString) const
 {
+    if (!m_replacementURLStringsForCSSStyleSheet.isEmpty() && is<HTMLLinkElement>(element)) {
+        if (auto* cssStyleSheet = downcast<HTMLLinkElement>(element).sheet()) {
```
RefPtr

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -539,6 +541,81 @@ static void addSubresourcesForAttachmentElementsIfNecessary(LocalFrame& frame, c
+    if (subresourcesDirectoryName.isEmpty())
+        return { };
+
+    auto document = frame.protectedDocument();
```
`RefPtr document = frame.document()` is the usual style.

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -539,6 +541,81 @@ static void addSubresourcesForAttachmentElementsIfNecessary(LocalFrame& frame, c
+
+    HashMap<RefPtr<CSSStyleSheet>, String> uniqueCSSStyleSheets;
+    HashMap<RefPtr<CSSStyleSheet>, String> relativeUniqueCSSStyleSheets;
+    auto& documentStyleSheets = document->styleSheets();
```
`auto&` -> `Ref`

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -539,6 +541,81 @@ static void addSubresourcesForAttachmentElementsIfNecessary(LocalFrame& frame, c
+    auto& documentStyleSheets = document->styleSheets();
+    for (unsigned index = 0; index < documentStyleSheets.length(); ++index) {
+        RefPtr styleSheet = documentStyleSheets.item(index);
+        if (!styleSheet || !is<CSSStyleSheet>(styleSheet.get()))
```
Just this will suffice:
```
if (!is<CSSStyleSheet>(styleSheet))
```

is<>() supports RefPtr and will do the null check for you. Your code was doing the null check twice.

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -539,6 +541,81 @@ static void addSubresourcesForAttachmentElementsIfNecessary(LocalFrame& frame, c
+        if (!styleSheet || !is<CSSStyleSheet>(styleSheet.get()))
+            continue;
+
+        CSSStyleSheet& cssStyleSheet = downcast<CSSStyleSheet>(*styleSheet);
```
Ref

szewai replied:
styleSheet is already a `RefPtr` (line 557), so it seems unnecessary to make cssStyleSheet Ref?

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -539,6 +541,81 @@ static void addSubresourcesForAttachmentElementsIfNecessary(LocalFrame& frame, c
+            String subresourceFileName = generateValidFileName(url, uniqueFileNames);
+            uniqueFileNames.add(subresourceFileName);
+            String subresourceFilePath = FileSystem::pathByAppendingComponent(subresourcesDirectoryName, subresourceFileName);
+            addResult.iterator->value = frame.isMainFrame() ? subresourceFilePath : subresourceFileName;
```
`addResult.iterator` is an iterator inside `uniqueCSSStyleSheets`. At line 588, you modified `uniqueCSSStyleSheets` so you cannot rely on your iterator still being valid here.

szewai replied:
line 588 modifies `uniqueSubresources`

rniwa commented on Source/WebCore/css/CSSStyleRule.cpp:
```diff
@@ -313,4 +313,10 @@ CSSRuleList& CSSStyleRule::cssRules() const
 
+void CSSStyleRule::getChildStyleSheets(HashSet<RefPtr<CSSStyleSheet>>& childStyleSheets)
+{
+    for (unsigned index = 0 ; index < nestedRules().size(); ++index)
```
Nit: there is a space before ;

rniwa commented on Source/WebCore/css/CSSStyleRule.cpp:
```diff
@@ -173,15 +173,15 @@ String CSSStyleRule::cssTextWithReplacementURLs(const HashMap<String, String>& r
 {
-    for (unsigned index = 0 ; index < nestedRules().size() ; index++)
-        rules.append("\n  ", item(index)->cssTextWithReplacementURLs(replacementURLStrings));
+    for (unsigned index = 0 ; index < nestedRules().size(); index++)
```
Nit: there is a space before ;

szewai committed:
Save style sheet resources with same URL in different files
https://bugs.webkit.org/show_bug.cgi?id=264520
rdar://118199671

Reviewed by Ryosuke Niwa.

In current implementation, if two link elements have the same url, we only create one file for storing the resource.
However, two link elements corresponds to two CSSStyleSheet objects, which can be modified independently by script. To
ensure the changes from script are captured, now we store content of CSSStyleSheet object in its own file.

API tests: WebArchive.SaveResourcesLink
           WebArchive.SaveResourcesLinksWithSameURL
           WebArchive.SaveResourcesCSSImportRule

* Source/WebCore/css/CSSFontFaceRule.cpp:
(WebCore::CSSFontFaceRule::cssTextWithReplacementURLs const):
* Source/WebCore/css/CSSFontFaceRule.h:
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::cssTextForDeclsAndRules const):
* Source/WebCore/css/CSSImportRule.cpp:
(WebCore::CSSImportRule::cssTextInternal const):
(WebCore::CSSImportRule::cssText const):
(WebCore::CSSImportRule::cssTextWithReplacementURLs const):
(WebCore::CSSImportRule::getChildStyleSheets):
* Source/WebCore/css/CSSImportRule.h:
* Source/WebCore/css/CSSRule.h:
(WebCore::CSSRule::cssTextWithReplacementURLs const):
(WebCore::CSSRule::getChildStyleSheets):
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::cssTextForRules const):
(WebCore::CSSStyleRule::cssTextWithReplacementURLs const):
(WebCore::CSSStyleRule::cssTextForRulesWithReplacementURLs const):
(WebCore::CSSStyleRule::getChildStyleSheets):
* Source/WebCore/css/CSSStyleRule.h:
* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::cssTextWithReplacementURLs):
(WebCore::CSSStyleSheet::getChildStyleSheets):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::MarkupAccumulator):
(WebCore::MarkupAccumulator::appendContentsForNode):
(WebCore::MarkupAccumulator::resolveURLIfNeeded const):
* Source/WebCore/editing/MarkupAccumulator.h:
(WebCore::MarkupAccumulator::MarkupAccumulator):
* Source/WebCore/editing/markup.cpp:
(WebCore::serializeFragment):
* Source/WebCore/editing/markup.h:
(WebCore::serializeFragment):
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::styleSheetContentWithReplacementURLs const): Deleted.
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLStyleElement.cpp:
(WebCore::HTMLStyleElement::textContentWithReplacementURLs const): Deleted.
* Source/WebCore/html/HTMLStyleElement.h:
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
(WebCore::addSubresourcesForCSSStyleSheetsIfNecessary):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: https://commits.webkit.org/270562@main
